Source/Core/Core/PowerPC/Jit64/Jit_LoadStore.cpp

```diff
@@ -507,7 +507,7 @@ void Jit64::stX(UGeckoInstruction inst)
       }
       else
       {
-        RCOpArg Ra = gpr.UseNoImm(a, RCMode::Write);
+        RCOpArg Ra = gpr.UseNoImm(a, RCMode::ReadWrite);
```

Comment:
This seems more like a workaround than a proper fix. Would keeping it as `RCMode::Write` work if you put the `MemoryExceptionCheck` call first? I'm assuming that the problem is that setting this as `RCMode::Write` causes the register allocation to think that the old value (the immediate value) is no longer valid.

Reply:
Ah, but wait... Jit64 isn't like JitArm64. It doesn't emit the code for flushing registers and exiting as soon as you call `MemoryExceptionCheck`, it does it later. I suppose we might need something different then? (Is this what `RevertableBind` is for?)

Reply:
After studying the code further, yes, I think `RevertableBind` is what we should be using here. Please try `RevertableBind(a, RCMode::Write)` and see if it works. If it does work, maybe we should have a `RevertableUseNoImm` function in addition to `RevertableBind`...?

More specifically, `a` should use `RevertableBind` when we have `update && jo.memcheck` for any loadstore. But when the `RCMode` for `a` is `ReadWrite`, we can technically skip it.

Reply:
Yes, that works properly (or at least RS2 does not crash with that when it did crash on master). I've changed it.

Unfortunately I don't really understand what is going on here.

Reply:
`RevertableBind` was designed for the following situation:

Imagine we're emulating a regular load instruction. There's nothing complicated going on in the address calculation, and we don't have update the index register. We just need to load a value from memory and store it to a register. In the general case, this is relatively straightforward. We allocate a register using `gpr.Bind(d, RCMode::Write)`, we load into that register, and that's it. (Of course, there's also a bunch of stuff going on to handle MMIO and such, but that's outside the scope of what we're discussing here.)

Now, let's imagine that `jo.memcheck` is set. This means that the load can either succeed, as before, or fail, in which case we need to bail out while keeping the state of every guest register the same as it was before. This is currently implemented in Jit64 as follows.

First, instead of using `gpr.Bind(d, RCMode::Write)`, we use `gpr.RevertableBind(d, RCMode::Write)`:

https://github.com/dolphin-emu/dolphin/blob/0a88c2329aad5636cfb36d63d7ebf2314d82344d/Source/Core/Core/PowerPC/Jit64/Jit_LoadStore.cpp#L137

When we do this, the register cache stores the current value of the guest register to `ppcState` (unless it already was in `ppcState`), ensures that a host register is allocated for the guest register, and marks the guest register as "revertable": https://github.com/dolphin-emu/dolphin/blob/d6437b7e461af0ef457c05be996d6fb30d7922ac/Source/Core/Core/PowerPC/Jit64/RegCache/JitRegCache.cpp#L733-L739

Later on, if we are bailing out because the load failed, Jit64 will write all registers to `ppcState` *except* the register(s) that we previously marked as revertable:

https://github.com/dolphin-emu/dolphin/blob/450ca0b69ea70f56e25670040f1dfdc01bf52ba9/Source/Core/Core/PowerPC/Jit64/Jit.cpp#L1077-L1083

What's going on with in the problem you have is really the same as in the problem I just described, just with the index register instead of the destination register. So I think it would make sense to reuse the solution we already have for this problem. It's not a perfect solution – the codegen in your initial version of this PR is in a sense better – but I would prefer to use the semantically correct solution here instead of using `ReadWrite`. We can always improve the "revertable" system later, and then that improvement will apply both to the destination register and the index register.